msamp/nn/distributed.py

```diff
@@ -0,0 +1,260 @@
+
+        return hook_fn
+
+    def _reduce_bucket(self, bucket_id):
+        """Reduce the gradients of all the parameters in a bucket.
+
+        Args:
+            bucket_id (int): The id of the bucket.
+        """
+        self.reduction_stream.wait_stream(torch.cuda.current_stream())
+        with torch.cuda.stream(self.reduction_stream):
+            # step 1: collect the gradients
+            param_ids = self.bucket_to_param_ids[bucket_id]
+            params = [self.parameters[i] for i in param_ids]
+            grads = [p.grad for p in params]
+            metas = [g.meta for g in grads]
+
+            # step 2: synchronize the amax
+            amaxs = torch.stack([g.abs().max().float() for g in grads])
+            scales = torch.stack([meta.scale for meta in metas])
+
+            amaxs.nan_to_num_(nan=torch.inf, posinf=torch.inf)    # convert NAN to INF since reduce ignores NAN
+            dist.all_reduce(amaxs, op=dist.ReduceOp.MAX, group=self.process_group)
+
+            # step 3: re-compute scaling factor and update meta.amax
+            wgrad_qtype = Dtypes.kfloat8_e4m3
+            fp_max = Floating.qfp_max[wgrad_qtype]
+            world_size = dist.get_world_size(self.process_group)
+            pre_scale = 1.0 / math.sqrt(world_size)
+            sf = ScalingMeta.compute_scaling_factor(amaxs, scales, fp_max, margin=0)
+            sf.mul_(pre_scale)
+            for meta, amax, scale in zip(metas, amaxs, sf):
+                meta.amax[0] = amax
+                meta.scale.copy_(scale)
+
+            # step 4: quantize the gradients to FP8
+            bucket_range = self.bucket_to_range[bucket_id]
+            bucket_start, bucket_end = bucket_range
+            bucket_offset = bucket_start
+            dummy_amax = torch.empty((1, ), dtype=torch.float32, device=self.device)
+            if self.buffer is None:
+                self.buffer = self._create_buffer()
+            for i, (grad, meta) in enumerate(zip(grads, metas)):
+                fp8_grad = TransformerEngineWrapper.cast_to_fp8(
+                    grad.view(1, -1),
+                    meta.scale,
+                    dummy_amax,
+                    meta.scale_inv,
+                    meta.qtype,
+                ).view_as(grad)
+                meta.scale_inv.data.copy_(torch.reciprocal(meta.scale))
+                grads[i] = None
+                # copy fp8_grad to buffer
+                grad_numel = grad.numel()
+                buf = self.buffer.narrow(0, bucket_offset, grad_numel).view_as(grad)
+                buf.copy_(fp8_grad)
+                params[i].grad = ScalingTensor(buf, meta)
+                params[i].grad.div_(world_size)
+                bucket_offset += grad_numel
+
+            flat_fp8_grads = self.buffer.narrow(0, bucket_start, bucket_end - bucket_start)
+
+            # step 5: allreduce the gradients
+            torch.cuda.default_stream().wait_stream(self.reduction_stream)
+            FP8Op.enable_fp8(wgrad_qtype)
```

Comment:
do you want to rename FP8Op to distributed related names like distOp?

Reply:
Yes, it is in another PR(https://github.com/Azure/MS-AMP/pull/94). Will replace it with DistOp in that PR after this PR merged.